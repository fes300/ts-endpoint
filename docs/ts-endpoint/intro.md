---
id: intro
title: setup
sidebar_label: setting up `ts-endpoint`
---

### install

```sh
# add ts-endpoint dep
$ pnpm add ts-endpoint
```

### description

`ts-endpoint` gives you a consistent way of defining every aspect of an `HTTP` endpoint from both the client (whose payload is the endpoint "Input") and the server (whose returning value is the endpoint "Output")
