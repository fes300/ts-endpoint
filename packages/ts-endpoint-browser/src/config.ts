

type HTTPProtocol = "http";
type HTTPSProtocol = "https";

export type HTTPClientProtocol = HTTPProtocol | HTTPSProtocol;

export type HTTPClientConfig = {
  protocol: HTTPClientProtocol;
  host: string;
  port?: number | undefined;
};
