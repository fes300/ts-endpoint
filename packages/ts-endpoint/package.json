{
  "name": "ts-endpoint",
  "version": "2.0.1-alpha.0",
  "description": "Simple endpoint definition for fullstack safety.",
  "main": "index.js",
  "repository": "https://github.com/fes300/ts-endpoint/tree/master/packages/ts-endpoint",
  "bugs": {
    "url": "https://github.com/fes300/ts-endpoint/issues?q=is%3Aopen+is%3Aissue+project%3Afes300%2Fts-endpoint%2F3"
  },
  "author": "Federico Sordillo <federicosordillo@gmail.com>",
  "license": "ISC",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "npm-publish-beta": "if [ -f lib/package.json ] ; then npm publish ./lib --tag beta; else echo echo \"no need to publish\"; fi",
    "npm-publish": "if [ -f lib/package.json ] ; then npm publish ./lib; else echo echo \"no need to publish\"; fi",
    "postversion": "cp package.json lib && cp -r lib lib2 && mv lib2 lib/lib",
    "build": "rm -rf ./lib && tsc",
    "type-test": "jest --config=jest.typetests-config.js",
    "runtime-test": "jest --config=jest.config.js",
    "test": "pnpm type-test && pnpm runtime-test"
  },
  "devDependencies": {
    "io-ts": "^2.2.20"
  },
  "dependencies": {
    "fp-ts": "^2.13.1",
    "ts-io-error": "workspace:*",
    "typelevel-ts": "^0.4.0"
  }
}
