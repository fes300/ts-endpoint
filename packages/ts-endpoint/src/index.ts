export * from './Endpoint';
export * from './helpers';
export * from './ResourceEndpoint';

