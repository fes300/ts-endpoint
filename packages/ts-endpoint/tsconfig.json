{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "lib": ["esnext"],
    "rootDir": "./src",
    "outDir": "./lib",
    "tsBuildInfoFile": "lib/.tsbuildinfo",
    "esModuleInterop": true
  },
  "include": ["src"]
}
