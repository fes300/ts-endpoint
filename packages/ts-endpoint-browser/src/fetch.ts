import { findFirst } from 'fp-ts/Array';
import * as E from 'fp-ts/Either';
import { toArray } from 'fp-ts/lib/Record';
import * as O from 'fp-ts/Option';
import { pipe } from 'fp-ts/pipeable';
import * as TA from 'fp-ts/TaskEither';
import { PathReporter } from 'io-ts/PathReporter';
import qs from 'qs';
import { EndpointErrors, MinimalEndpointInstance } from 'ts-endpoint';
import { Codec, IOError, NetworkErrorStatus } from 'ts-io-error';
import { HTTPClientConfig, StaticHTTPClientConfig } from './config';
import { FetchClient, GetHTTPClient, GetHTTPClientOptions } from './GetHTTPClient';

declare module './HKT' {
  interface URItoKind<A> {
    IOError: A extends Record<string, Codec<any, any, any>> ? IOError<A> : never;
  }
}

export const GetFetchHTTPClient = <A extends { [key: string]: MinimalEndpointInstance }>(
  config: HTTPClientConfig | StaticHTTPClientConfig,
  endpoints: A,
  options?: GetHTTPClientOptions
) => GetHTTPClient(config, endpoints, useBrowserFetch, options);

const getResponseJson = (r: Response, ignoreNonJSONResponse: boolean) => {
  if (r.body === null) {
    TA.left(undefined);
  }
  const jsonResponse = TA.tryCatch(
    () => r.json(),
    () =>
      new IOError(r.statusText, {
        status: r.status.toString(),
        kind: 'ServerError',
        meta: { message: 'response is not a json.' },
      })
  );

  return ignoreNonJSONResponse
    ? pipe(
        jsonResponse,
        TA.alt(() => TA.of(undefined))
      )
    : jsonResponse;
};

export const useBrowserFetch = <E extends MinimalEndpointInstance>(
  baseURL: string,
  e: E,
  options?: GetHTTPClientOptions
): FetchClient<E, 'IOError'> => {
  return (i: any) => {
    const path = `${baseURL}${e.getPath(i?.Params)}${i?.Query ? `?${qs.stringify(i?.Query)}` : ''}`;

    const body = i?.Body;
    const headers = { ...(i?.Headers ?? {}), ...options?.defaultHeaders };

    const response = pipe(
      TA.tryCatch(
        () =>
          fetch(path, {
            ...{
              headers,
              method: e.Method,
            },
            ...(i?.Body !== undefined ? { body: JSON.stringify(body) } : {}),
          }),
        () => new IOError('Network Error', { kind: 'NetworkError', status: NetworkErrorStatus })
      ),
      TA.chain((r: Response) => {
        const responseJson: TA.TaskEither<any, any> = TA.tryCatch(
          () => (r.body === null ? Promise.resolve(undefined) : r.json()),
          () => {
            return new IOError(r.statusText, {
              status: r.status.toString(),
              kind: 'ServerError',
              meta: { message: 'response is not a json.' },
            });
          }
        );

        const responseJsonWithDefault = options?.ignoreNonJSONResponse
          ? pipe(
              responseJson,
              TA.alt(() => {
                return TA.of(undefined);
              })
            )
          : responseJson;

        if (!r.ok) {
          if (e.Errors !== undefined) {
            const actualKnownError = pipe(
              e.Errors as EndpointErrors<string, Codec<any, any, any>>,
              toArray,
              findFirst(([status]) => {
                return status === r.status.toString();
              })
            );

            if (O.isSome(actualKnownError)) {
              const parsedKnownError = pipe(
                responseJsonWithDefault,
                TA.map((body) => actualKnownError.value[1].decode(body)),
                TA.chain((validation) => {
                  return pipe(
                    TA.fromEither(validation),
                    TA.fold(
                      (errors: unknown[]) => {
                        return TA.left(
                          new IOError(`Error decoding server KnownError response: ${errors}`, {
                            kind: 'DecodingError',
                            errors,
                          })
                        );
                      },

                      (parsedError) => {
                        return TA.left(
                          new IOError<any>(r.statusText, {
                            kind: 'KnownError',
                            status: r.status.toString(),
                            body: parsedError,
                          })
                        );
                      }
                    )
                  );
                })
              );

              return parsedKnownError;
            }
          }

          if (r.status >= 400 && r.status <= 451) {
            return pipe(
              getResponseJson(r, options?.ignoreNonJSONResponse ?? false),
              TA.chain((meta) => {
                return TA.left(
                  new IOError(r.statusText, {
                    kind: 'ClientError',
                    meta,
                    status: r.status.toString(),
                  })
                );
              })
            );
          } else {
            return pipe(
              getResponseJson(r, options?.ignoreNonJSONResponse ?? false),
              TA.chain((meta) => {
                return TA.left(
                  new IOError(r.statusText, {
                    kind: 'ServerError',
                    meta,
                    status: r.status.toString(),
                  })
                );
              })
            );
          }
        }

        const res = pipe(
          responseJsonWithDefault,
          TA.chain((json) => {
            return pipe(
              options?.mapInput ? options.mapInput(json) : json,
              e.Output.decode,
              TA.fromEither,
              TA.mapLeft((errors) => {
                return new IOError(
                  `Error decoding server response: ${PathReporter.report(E.left(errors))}`,
                  {
                    kind: 'DecodingError',
                    errors,
                  }
                );
              })
            );
          })
        );

        return res;
      }),
      TA.mapLeft((err) => {
        if (options?.handleError !== undefined) {
          return options.handleError(err, e);
        }
        return err;
      })
    );

    return response;
  };
};
