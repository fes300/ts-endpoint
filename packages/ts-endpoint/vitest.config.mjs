import { defineProject, mergeConfig } from 'vitest/config';
import vitestBaseConfig from '../../vitest.config.mjs';

export default mergeConfig(
  vitestBaseConfig,
  defineProject({
    test: {
      name: 'ts-endpoint',
      typecheck: { enabled: true },
    },
  })
);
