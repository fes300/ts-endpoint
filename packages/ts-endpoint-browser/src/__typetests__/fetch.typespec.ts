import { GetFetchHTTPClient } from '../fetch';
import { Endpoint } from 'ts-endpoint/lib';
import { StaticHTTPClientConfig } from '../config';
import * as t from 'io-ts';
import { InferFetchResult } from '..';
import { right } from 'fp-ts/Either';
import * as TA from 'fp-ts/TaskEither';
import { pipe } from 'fp-ts/pipeable';

const options: StaticHTTPClientConfig = {
  protocol: 'http',
  host: 'http://test',
  port: 2020,
};

const endpoints = {
  dummy: Endpoint({
    Input: {
      Query: t.type({ color: t.string }),
      Params: t.type({ id: t.string }),
    },
    Method: 'GET',
    getPath: ({ id }) => `users/${id}/crayons`,
    Output: t.type({ crayons: t.array(t.string) }),
  }),
  dummyNoParam: Endpoint({
    Input: {
      Query: t.type({ color: t.string }),
    },
    Method: 'GET',
    getPath: () => `users/crayons`,
    Output: t.type({ crayons: t.array(t.string) }),
  }),
  dummyWithoutInput: Endpoint({
    Method: 'GET',
    getPath: () => `users`,
    Output: t.type({ noInput: t.array(t.string) }),
  }),
  dummyWithError: Endpoint({
    Input: {
      Query: t.type({ color: t.string }),
      Params: t.type({ id: t.string }),
    },
    Errors: {
      401: t.type({ foo: t.string }),
      402: t.type({ baz: t.string }),
    },
    Method: 'GET',
    getPath: ({ id }) => `users/${id}/crayons`,
    Output: t.type({ crayons: t.array(t.string) }),
  }),
};

const fetchClient = GetFetchHTTPClient(options, endpoints, {
  handleError: (err, e) => err,
});

// @dts-jest:pass:snap should allow empty calls when input is not defined
fetchClient.dummyWithoutInput();

// @dts-jest:fail:snap should not allow calls with Body when input is not defined
fetchClient.dummyWithoutInput({
  Body: 1,
});

// @dts-jest:fail:snap should not allow calls with Query when input is not defined
fetchClient.dummyWithoutInput({
  Query: 1,
});

// @dts-jest:fail:snap should not allow calls with Params when input is not defined
fetchClient.dummyWithoutInput({
  Params: 1,
});

// @dts-jest:fail:snap should not allow calls with Header when input is not defined
fetchClient.dummyWithoutInput({
  Headers: 1,
});

// @dts-jest:fail:snap should not allow empty calls
fetchClient.dummy();

// @dts-jest:fail:snap should not allow empty-object as input
fetchClient.dummy({});

// @dts-jest:fail:snap should not allow calls not specifing Query
fetchClient.dummy({
  Params: { id: '123' },
});

fetchClient.dummy({
  // @dts-jest:fail:snap should not allow calls specifing Params not declared in the endpoint
  Params: { id: '123', foo: 'baz' },
  Query: { color: 'marrone' },
});

fetchClient.dummy({
  Params: { id: '123' },
  Query: { color: 'marrone' },
  // @dts-jest:fail:snap should not allow to add Body when not declared in the endpoint
  Body: { foo: 'baz' },
});

// @dts-jest:fail:snap InferFetchResult should return the type of the Endpoint.Output
const dummyResultWrong: InferFetchResult<typeof fetchClient.dummy> = right({});

// @dts-jest:pass:snap InferFetchResult should return the type of the Endpoint.Output
const dummyResultCorrect: InferFetchResult<typeof fetchClient.dummy> = right({
  crayons: ['brown'],
});

pipe(
  fetchClient.dummyWithError({
    Params: { id: '123' },
    Query: { color: 'marrone' },
  }),
  TA.mapLeft((err) => {
    if (err.details.kind === 'KnownError') {
      if (err.details.status === 401) {
        // @dts-jest:pass:snap you can access KnownErrors with the correct typeguard
        err.details.body.foo;

        // @dts-jest:fail:snap you cannot access KnownErrors without the correct typeguard
        err.details.body.baz;
      }
    }
  })
);
