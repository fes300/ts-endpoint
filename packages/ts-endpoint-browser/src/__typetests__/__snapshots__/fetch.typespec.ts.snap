// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InferFetchResult should return the type of the Endpoint.Output (type) should match snapshot 1`] = `
"Type 'Either<IOError<never>, {}>' is not assignable to type 'Either<IOError<never>, { crayons: string[]; }>'.
  Type 'Right<{}>' is not assignable to type 'Either<IOError<never>, { crayons: string[]; }>'.
    Type 'Right<{}>' is not assignable to type 'Right<{ crayons: string[]; }>'.
      Property 'crayons' is missing in type '{}' but required in type '{ crayons: string[]; }'."
`;

exports[`InferFetchResult should return the type of the Endpoint.Output (type) should match snapshot 2`] = `"any"`;

exports[`should allow empty calls when input is not defined (type) should match snapshot 1`] = `"TaskEither<IOError<never>, { noInput: string[]; }>"`;

exports[`should not allow calls not specifing Query (type) should match snapshot 1`] = `
"Argument of type '{ Params: { id: string; }; }' is not assignable to parameter of type '{ Params: { id: string; }; Headers?: never; Query: { color: string; }; Body?: never; }'.
  Property 'Query' is missing in type '{ Params: { id: string; }; }' but required in type '{ Params: { id: string; }; Headers?: never; Query: { color: string; }; Body?: never; }'."
`;

exports[`should not allow calls specifing Params not declared in the endpoint (type) should match snapshot 1`] = `
"Type '{ id: string; foo: string; }' is not assignable to type '{ id: string; }'.
  Object literal may only specify known properties, and 'foo' does not exist in type '{ id: string; }'."
`;

exports[`should not allow calls with Body when input is not defined (type) should match snapshot 1`] = `"Argument of type '{ Body: number; }' is not assignable to parameter of type 'void'."`;

exports[`should not allow calls with Header when input is not defined (type) should match snapshot 1`] = `"Argument of type '{ Headers: number; }' is not assignable to parameter of type 'void'."`;

exports[`should not allow calls with Params when input is not defined (type) should match snapshot 1`] = `"Argument of type '{ Params: number; }' is not assignable to parameter of type 'void'."`;

exports[`should not allow calls with Query when input is not defined (type) should match snapshot 1`] = `"Argument of type '{ Query: number; }' is not assignable to parameter of type 'void'."`;

exports[`should not allow empty calls (type) should match snapshot 1`] = `"Expected 1 arguments, but got 0."`;

exports[`should not allow empty-object as input (type) should match snapshot 1`] = `
"Argument of type '{}' is not assignable to parameter of type '{ Params: { id: string; }; Headers?: never; Query: { color: string; }; Body?: never; }'.
  Type '{}' is missing the following properties from type '{ Params: { id: string; }; Headers?: never; Query: { color: string; }; Body?: never; }': Params, Query"
`;

exports[`should not allow to add Body when not declared in the endpoint (type) should match snapshot 1`] = `"Type '{ foo: string; }' is not assignable to type 'never'."`;

exports[`you can access KnownErrors with the correct typeguard (type) should match snapshot 1`] = `"string"`;

exports[`you cannot access KnownErrors without the correct typeguard (type) should match snapshot 1`] = `"Property 'baz' does not exist on type '{ foo: string; }'."`;
